pkg/parser/tac/parse.go

```diff
@@ -0,0 +1,183 @@
+// 2. The second line of the file are the columns that will be a 1:1 match to the TransportationAccountingCodeTrdmFileRecord struct in pipe delimited format.
+// 3. There are 23 values per line, excluding the security classification. Again, to know what these values are refer to note #2.
+// 4. All values are in pipe delimited format.
+// 5. Null values will be present, but are not acceptable for TRNSPRTN_ACNT_CD.
+func Parse(file io.Reader) ([]models.TransportationAccountingCodeDesiredFromTRDM, error) {
+
+	// Init variables
+	var codes []models.TransportationAccountingCodeDesiredFromTRDM
+	scanner := bufio.NewScanner(file)
+	var columnHeaders []string
+
+	// Skip first line as it does not hold any necessary data for parsing.
+	// Additionally, this will check if it is empty.
+	if !scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+		return nil, errors.New("empty file")
+	}
+
+	// Read the second line to get the column names from the TRDM file, verify them,
+	// and then proceed with parsing the rest of the file.
+	if scanner.Scan() {
+		columnHeaders = strings.Split(scanner.Text(), "|")
+		ensureFileStructMatchesColumnNames(columnHeaders)
+	}
+
+	// Process the lines of the .txt file into modeled codes
+	codes, err := processLines(scanner, columnHeaders, codes)
+	if err != nil {
+		return nil, err
+	}
+
+	return codes, nil
+}
+
+// Compare a struct's field names to the columns retrieved from the .txt file
+func ensureFileStructMatchesColumnNames(columnNames []string) error {
+	if len(columnNames) == 0 {
+		return errors.New("column names were not parsed properly from the second line of the tac file")
+	}
+	expectedColumnNames := getFieldNames(models.TransportationAccountingCodeTrdmFileRecord{})
+	if !reflect.DeepEqual(columnNames, expectedColumnNames) {
+		return errors.New("column names parsed do not match the expected format of tac file records")
+	}
+	return nil
+}
+
+// This function gathers the struct field names for comparison to
+// line 2 of the .txt file - The columns
+func getFieldNames(obj interface{}) []string {
+	var fieldNames []string
+
+	t := reflect.TypeOf(obj)
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		fieldNames = append(fieldNames, field.Name)
+	}
+
+	return fieldNames
+}
+
+// Removes all TACs with an expiration date in the past
+func PruneExpiredTACsDesiredFromTRDM(codes []models.TransportationAccountingCodeDesiredFromTRDM) []models.TransportationAccountingCodeDesiredFromTRDM {
+	var pruned []models.TransportationAccountingCodeDesiredFromTRDM
+
+	// If the expiration date is not before time.Now(), then it is not expired and should be appended to the pruned array
+	for _, code := range codes {
+		if !code.ExpirationDate.Before(time.Now()) {
+			pruned = append(pruned, code)
+		}
+	}
+
+	return pruned
+}
+
+// Consolidates TACs with the same TAC value. Duplicate "Transaction", aka description, calues are combined with a delimeter of ". Additional description found: "
```

Comment:
Comment on line 63 "calues". should that say "values"?

Reply:
Yes it should, thank you for spotting that